source/agora/crypto/Bech32.d

```diff
@@ -0,0 +1,486 @@
+        // process. Simplifying:
+        // c'(x) = (f(x) * x + v_i) mod g(x)
+        //         ((f(x) mod g(x)) * x + v_i) mod g(x)
+        //         (c(x) * x + v_i) mod g(x)
+        // If c(x) = c0*x^5 + c1*x^4 + c2*x^3 + c3*x^2 + c4*x + c5,
+        // we want to compute
+        // c'(x) = (c0*x^5 + c1*x^4 + c2*x^3 + c3*x^2 + c4*x + c5) * x + v_i mod g(x)
+        //       = c0*x^6 + c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i mod g(x)
+        //       = c0*(x^6 mod g(x)) + c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i
+        // If we call (x^6 mod g(x)) = k(x), this can be written as
+        // c'(x) = (c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i) + c0*k(x)
+
+        // First, determine the value of c0:
+        ubyte c0 = c >> 25;
+
+        // Then compute c1*x^5 + c2*x^4 + c3*x^3 + c4*x^2 + c5*x + v_i:
+        c = ((c & 0x1ffffff) << 5) ^ v_i;
+
+        // Finally, for each set bit n in c0, conditionally add {2^n}k(x):
+        if (c0 & 1)  c ^= 0x3b6a57b2; //     k(x) = {29}x^5 + {22}x^4 + {20}x^3 +
+                                      //            {21}x^2 + {29}x + {18}
+        if (c0 & 2)  c ^= 0x26508e6d; //  {2}k(x) = {19}x^5 +  {5}x^4 +     x^3 +
+                                      //            {3}x^2 + {19}x + {13}
+        if (c0 & 4)  c ^= 0x1ea119fa; //  {4}k(x) = {15}x^5 + {10}x^4 +  {2}x^3 +
+                                      //            {6}x^2 + {15}x + {26}
+        if (c0 & 8)  c ^= 0x3d4233dd; //  {8}k(x) = {30}x^5 + {20}x^4 +  {4}x^3 +
+                                      //            {12}x^2 + {30}x + {29}
+        if (c0 & 16) c ^= 0x2a1462b3; // {16}k(x) = {21}x^5 +     x^4 +  {8}x^3 +
+                                      //            {24}x^2 + {21}x + {19}
+    }
+    return c;
+}
+
+/*******************************************************************************
+
+    Expand a HRP(Human Readable Part) for use in checksum computation.
+
+    Params:
+        hrp = human redable part
+
+    Returns:
+        5-bits values for the hrp
+
+*******************************************************************************/
+
+ubyte[] expand_hrp (const string hrp)
+{
+    ubyte[] ret;
+    ret.reserve(hrp.length + 90);
+    ret.length = hrp.length * 2 + 1;
+    for (size_t i = 0; i < hrp.length; ++i)
+    {
+        ubyte c = hrp[i];
+        ret[i] = c >> 5;
+        ret[i + hrp.length + 1] = c & 0x1f;
+    }
+    ret[hrp.length] = 0;
+    return ret;
+}
+
+/*******************************************************************************
+
+    Create a checksum
+
+    Params:
+        hrp = human redable part
+        values = values to get checksum from
+
+    Returns:
+        an checksum
+
+*******************************************************************************/
+
+ubyte[] create_checksum (string hrp, const ubyte[] values)
+{
+    ubyte[] enc = expand_hrp(hrp);
+    enc ~= cast(ubyte [])values;
+    enc.length = enc.length + 6;
+    uint mod = polymod(enc) ^ 1;
```

Comment:
Did you find a good resource about Bech32M ?

Reply:
https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki

Reply:
Thanks for the information.